npu.user.js

```diff
@@ -1060,24 +1076,33 @@ var npu = {
 						});
 						
 						if(markRows.size() > 0) {
-							lastMark = markRows.last();
+							var lastMark = markRows.last();
 							var mark = $("td:nth-child(4)", lastMark).text().trim();
-							var classToBeAdded = null;
-							
-							if(npu.isPassingGrade(mark)) {
-								classToBeAdded = "npu_completed";
+							var passed = npu.isPassingGrade(mark);
+							!subscribed && (classToBeAdded = passed ? "npu_completed" : "npu_failed");
+							if(passed) {
 								row.attr("data-completed", "1");
 								row.add(subRow)[filterEnabled ? "addClass" : "removeClass"]("npu_hidden");
 							}
-							else {
-								classToBeAdded = "npu_failed";
-							}
-							
-							if (classToBeAdded && !row.hasClass("gridrow_blue")) {
-								row.addClass(classToBeAdded);
-							}
+						}
+						
+						classToBeAdded = classToBeAdded || (subscribed ? "npu_subscribed" : "npu_found");
+						row.addClass(classToBeAdded);
+						
+						if(!$("#upFilter_cmbSubjects").val() || $("#upFilter_cmbSubjects").val() === "0") {
+							$.examSubjectFilterCache = $.examSubjectFilterCache || {};
+							$.examSubjectFilterCache[courseCode] = classToBeAdded;

```

Comment:
`$.examSubjectFilterCache` now has a much simpler structure where the keys are the subject codes and the corresponding values are the classes to be applied. This is made possible by the fact that we now parse the subject code in the `<option>` text and look it up directly (see below).